tests/test_software_upgrade.sh

```diff
@@ -0,0 +1,169 @@
+#!/bin/bash 
+# Test a gaia software upgrade via governance proposal.
+# It assumes gaia is running on the local host.
+
+gaia_host=$1
+gaia_port=$2
+upgrade_version=$3
+
+echo "Attempting upgrade to $upgrade_version."
+
+# Add gaiad to PATH
+echo "Adding gaiad to PATH..."
+export PATH="$PATH:~/.gaia/cosmovisor/current/bin"
+echo "PATH=$PATH"
+
+# Get the current gaia version from the API
+gaiad_version=$(curl -s http://$gaia_host:$gaia_port/abci_info | jq -r .result.response.version)
+
+# Submit upgrade proposal
+gaiad_version_major=${gaiad_version:1:1}
+echo "Current major version: $gaiad_version_major"
+
+upgrade_version_major=${upgrade_version:1:1}
+echo "Upgrade major version: $upgrade_version_major"
+major_difference=$[ $upgrade_version_major-$gaiad_version_major ]
+
+if [ $major_difference -eq 1 ]; then
+    if [ $upgrade_version_major -eq 7 ]; then
+        upgrade_name="v7-Theta"
+    elif [ $upgrade_version_major -eq 8 ]; then
+        upgrade_name="v8-Rho"
+    fi
+fi
+
+if [ -n "$upgrade_name" ]; then
+    echo "Upgrading to $upgrade_name."
+
+    # Set time to wait for proposal to pass
+    voting_period=$(curl -s http://localhost:26657/genesis\? | jq -r '.result.genesis.app_state.gov.voting_params.voting_period')
+    # voting_period=${voting_period::-1}
+    echo "Adding 5s to the voting period ($voting_period) to calculate the upgrade height."
+    voting_period_seconds=${voting_period::-1}
+    let voting_waiting_time=$voting_period_seconds+5
+
+    # Calculate upgrade height
+    let voting_blocks_delta=$voting_period_seconds/5+3
+    height=$(curl -s http://$gaia_host:$gaia_port/block | jq -r .result.block.header.height)
+    upgrade_height=$(($height+$voting_blocks_delta))
+    echo "Upgrade block height set to $upgrade_height."
+
+    # Get validator address and account sequence
+    # Sequence number must be added 1 for the vote to be accepted.
+    echo "Querying gaiad for the validator address and account sequence..."
+    validator_address=$(gaiad keys list --keyring-backend test --output json | jq -r '.[0].address')
+    validator_sequence=$[ $(gaiad query auth account $validator_address --output json | jq -r '.sequence')+1 ]
+    echo "The validator has address $validator_address, setting sequence # to $validator_sequence."
+
+    # Manual download
```

Comment:
should we clean this up?

Reply:
I had to get the validator sequence number to submit the proposal (see the next comment).
Or do you mean the manual download block? I thought we could maybe use it in the future in case we wanted to test it besides auto-download, but I'm ok with removing it.